include/picolibrary/circular_buffer.h

```diff
@@ -603,6 +604,464 @@ class Circular_Buffer<T, N, Without_Interrupt_Support, With_Overflow_Underflow_P
+ *            the interrupt may write data to the circular buffer.
+ *
+ * \tparam T The circular buffer element type.
+ * \tparam N The maximum number of elements in the circular buffer.
+ */
+template<typename T, std::size_t N>
+class Circular_Buffer<T, N, With_Interrupt_Support, Without_Overflow_Underflow_Protection> {
+  public:
+    static_assert( N > 0 );
+
+    /**
+     * \brief The circular buffer element type.
+     */
+    using Value = T;
+
+    /**
+     * \brief The number of elements in the circular buffer.
+     */
+    using Size = std::size_t;
+
+    /**
+     * \brief Constructor.
+     */
+    Circular_Buffer() noexcept = default;
+
+    Circular_Buffer( Circular_Buffer && ) = delete;
+
+    Circular_Buffer( Circular_Buffer const & ) = delete;
+
+    /**
+     * \brief Destructor.
+     */
+    ~Circular_Buffer() noexcept
+    {
+        clear();
+    }
+
+    auto operator=( Circular_Buffer && ) = delete;
+
+    auto operator=( Circular_Buffer const & ) = delete;
+
+    /**
+     * \brief Access the first element in the circular buffer.
+     *
+     * \warning Calling this function on an empty circular buffer results in undefined
+     *          behavior.
+     *
+     * \return The first element in the circular buffer.
+     */
+    auto front() noexcept -> Value volatile &
+    {
+        return *std::launder( reinterpret_cast<Value volatile *>( m_read ) );
+    }
+
+    /**
+     * \brief Access the first element in the circular buffer.
+     *
+     * \warning Calling this function on an empty circular buffer results in undefined
+     *          behavior.
+     *
+     * \return The first element in the circular buffer.
+     */
+    auto front() const noexcept -> Value const volatile &
+    {
+        return *std::launder( reinterpret_cast<Value const volatile *>( m_read ) );
+    }
+
+    /**
+     * \brief Access the last element in the circular buffer.
+     *
+     * \warning Calling this function on an empty circular buffer results in undefined
+     *          behavior.
+     *
+     * \return The last element in the circular buffer.
+     */
+    auto back() noexcept -> Value volatile &
+    {
+        Storage volatile * write;
+        {
+            auto guard = Interrupt_Guard<Restore_Interrupt_Enable_State>{};
+
+            write = m_write;
+        }
+
+        return *std::launder( reinterpret_cast<Value volatile *>(
+            write == &m_storage[ 0 ] ? &m_storage[ N - 1 ] : write - 1 ) );
+    }
+
+    /**
+     * \brief Access the last element in the circular buffer.
+     *
+     * \warning Calling this function on an empty circular buffer results in undefined
+     *          behavior.
+     *
+     * \return The last element in the circular buffer.
+     */
+    auto back_from_interrupt() noexcept -> Value volatile &
+    {
+        return *std::launder( reinterpret_cast<Value volatile *>(
+            m_write == &m_storage[ 0 ] ? &m_storage[ N - 1 ] : m_write - 1 ) );
```

Comment:
Making a non-volatile copy of `m_write` and then using the copy would allow for performance optimizations.
```suggestion
        auto write = Storage volatile * { m_write };
        
        return *std::launder( reinterpret_cast<Value volatile *>(
            write == &m_storage[ 0 ] ? &m_storage[ N - 1 ] : write - 1 ) );
```